### DISCUSSION TITLE
Is the TCP payload is recorded or complete IP packet is recorded..?

### DISCUSSION THREAD

**Discussion by charankamarapu** [mitmproxy/mitmproxy#7254] [Q&A]
I was looking into windows redirector, I have some doubts regarding the feature - Replay HTTP responses of a previously recorded server. As far as I understand .

1. Client creates tries to create a tcp/udp connection.
2. We intercept that call and send the buffer/InternetPacket to proxy.
3. The proxy reads the data and creates a connection to the destination.
4. The InternetPacket or buffer that is received is recorded/saved as request and then sent  to destination.
5. Then data/response is read from the server and then saved/recorded as response and then sent to redirector as InternetPacket and re-injected .

Basically the request and response is captured.

But what I don't understand is
- Is the TCP payload saved/recorded or the complete Internet packet along with header is saved/recorded?
- From the output of saved request and response, I can see that only the tcp payload is saved. So my question is when the response is replayed(Replay HTTP responses of a previously recorded server) only TCP payload is present. So how can I send the 
internetPacket to redirector because there is only tcp payload and not the complete IP packet.

I tried to read the code but couldn't find much. Can you please help me here..?

**Comment by mhils** [comment]
> Is the TCP payload saved/recorded or the complete Internet packet along with header is saved/recorded?

Neither. For local redirect mode on Windows,

1. Raw TCP packets are intercepted and fed into smoltcp.
2. smoltcp processes them and gives us raw TCP streams.
3. those cross the Rust<->Python boundary via [Stream.read](https://mitmproxy.github.io/mitmproxy_rs/mitmproxy_rs.html#Stream.read).
4. mitmproxy typically processes TCP streams into [HTTP flows](https://docs.mitmproxy.org/stable/api/mitmproxy/http.html#HTTPFlow), which may then be saved to disk.

See the illustration in https://github.com/mitmproxy/mitmproxy_rs/blob/main/README.md fir 1-3.

mitmproxy's client replay and server replay operate on flows only.

**Comment by charankamarapu** [comment]
its Raw IP packets right..? Not Raw TCP packets..?

**Reply by mhils** [reply]
Yes, correct. We intercept raw IP packets. TCP payloads are fed to smol, UDP payloads bypass that.

**Reply by charankamarapu** [reply]
In windows at the end, IP packets should be injected back right...? what would be the IP Header of those packets, because we are going to replay the recorded http flow which doesn't have any clue about IP Header.?

**Reply by mhils** [reply]
This just works the other way around: HTTPResponse is turned into raw bytes, they are put onto a TCP stream, and smoltcp turns that TCP stream into individual packets (and handles retransmissions etc).

**Reply by charankamarapu** [reply]
how will the smoltcp add the IP Header which will consists data other than the TCP payload ..? smoltcp should add it properly so that windows will accept and reinfect it and reach the client connection again right..? Any idea how it is done and from where smoltcp will get that data from..?

**Reply by mhils** [reply]
I suggest you read up on how TCP (replay) works. smoltcp is a full user-space networking stack.

**Comment by charankamarapu** [comment]
Can you provide any resources if possible..?
